pybedlite/overlap_detector.py

```diff
@@ -126,36 +177,94 @@ def from_bedrecord(cls: Type["Interval"], record: BedRecord) -> "Interval":
+      * `start`: A 0-based start position
+      * `end`: A 0-based exclusive end position
+
+    Interval-like Python objects may also contain strandedness information which will be used
+    for sorting them in :func:`~pybedlite.overlap_detector.OverlapDetector.get_overlaps` using
+    either of the following properties if they are present:
+
+      * `negative (bool)`: Whether or not the feature is negative stranded or not
+      * `strand (BedStrand)`: The BED strand of the feature
+      * `strand (str)`: The strand of the feature (`"-"` for negative)
 
     The same interval may be added multiple times, but only a single instance will be returned
-    when querying for overlaps.  Intervals with the same coordinates but different names are
-    treated as different intervals.
+    when querying for overlaps.
 
     This detector is the most efficient when all intervals are added ahead of time.
     """
 
-    def __init__(self) -> None:
+    def __init__(self, intervals: Optional[Iterable[_GenericGenomicSpan]] = None) -> None:
         # A mapping from the contig/chromosome name to the associated interval tree
         self._refname_to_tree: Dict[str, cr.cgranges] = {}  # type: ignore
         self._refname_to_indexed: Dict[str, bool] = {}
-        self._refname_to_intervals: Dict[str, List[Interval]] = {}
+        self._refname_to_intervals: Dict[str, List[_GenericGenomicSpan]] = {}
+        if intervals is not None:
+            self.add_all(intervals)
 
-    def __iter__(self) -> Iterator[Interval]:
+    def __iter__(self) -> Iterator[_GenericGenomicSpan]:
         """Iterates over the intervals in the overlap detector."""
         return itertools.chain(*self._refname_to_intervals.values())
 
-    def add(self, interval: Interval) -> None:
+    @staticmethod
+    def _reference_sequence_name(interval: GenomicSpan) -> str:
+        """Return the reference name of a given interval."""
+        if isinstance(interval, Interval) or hasattr(interval, "refname"):
+            return interval.refname
+        elif isinstance(interval, BedRecord) or hasattr(interval, "chrom"):
+            return interval.chrom
+        elif hasattr(interval, "contig"):
+            return interval.contig
+        else:
+            raise ValueError(
+                f"Genomic span is missing a reference sequence name property: {interval}"
+            )
+
+    @staticmethod
+    def _is_negative(interval: GenomicSpan) -> bool:
+        """Determine if this is a negative stranded interval or not."""
+        return (
+            (hasattr(interval, "negative") and interval.negative)
+            or (
+                hasattr(interval, "strand")
+                and isinstance(interval.strand, BedStrand)
+                and interval.strand is BedStrand.Negative
+            )
+            or (
+                hasattr(interval, "strand")
+                and isinstance(interval.strand, str)
+                and interval.strand == "-"
+            )
+        )
```

Comment:
Although these are pure functions and I could have implemented them in the package namespace, I made them class static methods because an advanced user (developer) may want to temporarily override them via inheritance.